### PR TITLE
[Tabular][fastai] Refactoring

### PR DESCRIPTION
**Pull request by gradientsky**
*Description of changes:*
- extracted and reused model-specific feature from tabular NN into fastai tabular NN
- refactoring of fastai code

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

### CODE REVIEW

File: tabular/src/autogluon/tabular/models/fastainn/metrics.py [hunk 1]
```diff
@@ -0,0 +1,79 @@
+import logging
+
+from autogluon.core.constants import REGRESSION
+
+logger = logging.getLogger(__name__)
+
+
+def get_objective_func_name(problem_type, stopping_metric):
```
Comment (gradientsky) [inline tabular/src/autogluon/tabular/models/fastainn/metrics.py:8]: These just moved into separate file to improve readability, no actual change

File: tabular/src/autogluon/tabular/models/tabular_nn/tabular_nn_model.py [hunk 2]
```diff
@@ -601,43 +574,6 @@ def _get_feature_type_map(self):
                 raise ValueError("unknown feature type encountered")
         return feature_type_map
 
-    def _create_preprocessor(self, impute_strategy, max_category_levels):
```
Comment (gradientsky) [inline tabular/src/autogluon/tabular/models/tabular_nn/tabular_nn_model.py:604]: just moved into separate file to enable reuse, no functional change

File: tabular/src/autogluon/tabular/models/tabular_nn/tabular_nn_preprocessor.py [hunk 3]
```diff
@@ -0,0 +1,95 @@
...
+
+from sklearn.compose import ColumnTransformer
+from sklearn.impute import SimpleImputer
+from sklearn.pipeline import Pipeline
+from sklearn.preprocessing import StandardScaler, QuantileTransformer, FunctionTransformer  # PowerTransformer
+
+from .categorical_encoders import OneHotMergeRaresHandleUnknownEncoder, OrdinalMergeRaresHandleUnknownEncoder
+
+
+class TabularNeuralNetPreprocessor(object):
```
Comment (gradientsky) [inline tabular/src/autogluon/tabular/models/tabular_nn/tabular_nn_preprocessor.py:11]: just moved into separate file to enable reuse, no functional change

**Comment by szha** [comment]
Job PR-829-1 is done. 
Docs are uploaded to http://autogluon-staging.s3-website-us-west-2.amazonaws.com/PR-829/1/index.html

**Review by Innixma — CHANGES_REQUESTED** [review]
Marking this as request changes to ensure it isn't merged prior to benchmarking.

- [ ] Benchmark on AutoMLBenchmark to ensure performance and stability are not degraded.

**Review by jwmueller — APPROVED** [review]
Code looks good to me. 

When benchmarking it could also be good to look at whether:
1) this change improves FastAI model
2) this change improves autogluon-ensemble (I'd have slight worry that sharing common preprocessing between MXNet and FastAI nets may reduce ensemble diversity, even if it does help FastAI nets individually)

**Comment by Innixma** [comment]
> Code looks good to me.
> 
> When benchmarking it could also be good to look at whether:
> 
> 1. this change improves FastAI model
> 2. this change improves autogluon-ensemble (I'd have slight worry that sharing common preprocessing between MXNet and FastAI nets may reduce ensemble diversity, even if it does help FastAI nets individually)

I also feel that the ensemble score might drop if the gains are only minor for FastAI, so it will be interesting to see the results.

**Pull request closed without merging**